### PR TITLE
Revamped RequestData on DgsContext, support for @RequestHeader and @InputArgument without name

### PR DESCRIPTION
**Pull request by paulbakker**
Improves access to request data. This should always be available, also when custom context is provided. Also added `@RequestHeader` and improved support for annotated arguments without names such as `@RequestHeader` or `@InputArgument`.

Note that the change to how RequestData is stored is potentially breaking, although the old API wasn't advertised. To prevent this the old mechanism is still available but will show as deprecated.

This PR fixes: 
* #93
* #67 
* #95

These changes were combined in a single PR because they are related to the same code.

New features
====

-  The signature of `DgsContext` is now `DgsContext(val customContext: Any?, val requestData: DgsRequestData?)`
- Previously the RequestData was set as the custom context object. If a custom context was user-provided (which is common) the RequestData would not be set at all, and there was no good way to add it manually to the custom context. 
- `DgsContext.getRequestData(dataFetchtingEnvironment` as shortcut to get RequestData
- DgsRequestData also has a reference to WebRequest
- `@RequestHeader` as a shortcut to get a specific header value.
- Headers are technically Lists, so the annotation can be either used on a `List` or on a `String`. In the latter case multiple values will be concatinated.
- `@InputArgument` without specifying a name explicitly. If no name is provided the argument name is discovered using https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/DefaultParameterNameDiscoverer.html

### CODE REVIEW

**Review by srinivasankavitha — APPROVED** [review]

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsDataFetchingEnvironment.kt [hunk 1]
```diff
@@ -34,6 +35,15 @@ import java.util.*
 
 class DgsDataFetchingEnvironment(private val dfe: DataFetchingEnvironment) : DataFetchingEnvironment {
 
+    fun getDgsContext(): DgsContext {
+        val context = dfe.getContext<Any>()
+        if (context is DgsContext) {
+            return context
+        } else {
+            throw RuntimeException("Context object is not a DgsContext. This method does not work if you have a custom implementation of DgsContextBuilder")
```
Comment (berngp) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsDataFetchingEnvironment.kt:43]: Show the type of the `context` in the exception.

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt [hunk 2]
```diff
@@ -241,13 +244,13 @@ class DgsSchemaProvider(private val applicationContext: ApplicationContext,
     private fun invokeDataFetcher(method: Method, dgsComponent: Any, environment: DataFetchingEnvironment): Any? {
         return try {
             val args = mutableListOf<Any?>()
-
-            method.parameters.forEach { parameter ->
+            val parameterNames = DefaultParameterNameDiscoverer().getParameterNames(method) ?: emptyArray()
+            method.parameters.forEachIndexed { idx, parameter ->
```
Comment (berngp) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt:248]: the lambda is probably getting a bit large, should we reflect it as an internal method?

**Review by berngp — APPROVED** [review]
Great feature.

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt (hunk 2 above)
Comment (paulbakker) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt:248]: Planning to address that separately

**Pull request merged**